cmd/snap-confine/snap-confine.c

```diff
@@ -515,16 +541,54 @@ int main(int argc, char **argv)
 	return 1;
 }
 
-static void enter_classic_execution_environment(void)
+static void enter_classic_execution_environment(const sc_invocation * inv)
 {
 	/* 'classic confinement' is designed to run without the sandbox inside the
 	 * shared namespace. Specifically:
-	 * - snap-confine skips using the snap-specific mount namespace
+	 * - snap-confine skips using the snap-specific, private, mount namespace
 	 * - snap-confine skips using device cgroups
 	 * - snapd sets up a lenient AppArmor profile for snap-confine to use
 	 * - snapd sets up a lenient seccomp profile for snap-confine to use
 	 */
+
 	debug("skipping sandbox setup, classic confinement in use");
+
+	if (!sc_feature_enabled(SC_FEATURE_PARALLEL_INSTANCES)) {
+		return;
+	}
+	/* all of the following code is experimental */
+
+	debug
+	    ("(experimental) unsharing the mount namespace (per-classic-snap)");
+	if (unshare(CLONE_NEWNS) < 0) {
+		die("cannot unshare the mount namespace for parallel installed classic snap");
+	}
+
+	/* Parallel installed classic snap get special handling */
+	if (!sc_streq(inv->snap_instance, inv->snap_name)) {
+		/* Construct a mount namespace where the snap instance directories are
+		 * visible under the regular snap name. In order to do that we will:
+		 *
+		 * - convert SNAP_MOUNT_DIR into a mount point (global init)
+		 * - convert /var/snap into a mount point (global init)
+		 * - create a new mount namespace (done earlier)
```

Comment:
It is a little weird to have this big comment here with the 'create a new mount namespace (done earlier)' comment where that applies to the previous if conditional. Why not move this whole big comment above unshare() for clarity?